src/nvim/event/rstream.c

```diff
@@ -167,9 +169,19 @@ static void fread_idle_cb(uv_idle_t *handle)
   invoke_read_cb(stream, false);
 }
 
-static void invoke_read_cb(Stream *stream, bool eof)
+static void read_event(void **argv)
 {
+  Stream *stream = argv[0];
+  bool eof = (uintptr_t)argv[1];
   if (stream->read_cb) {
     stream->read_cb(stream, stream->buffer, stream->data, eof);
   }
+  if (eof && stream->internal_eof_cb) {
+    stream->internal_eof_cb(stream, stream->internal_data);
+  }
+}
+
+static void invoke_read_cb(Stream *stream, bool eof)
+{
+  queue_put(stream->events, read_event, 2, stream, (void *)(uintptr_t)eof);
```

Comment:
I have made a small change to pty processes: Now they only duplicate the master fd to stdout instead of stdout and stderr, and it seems to have fixed some random errors in terminal tests.

Great job spotting this bug, it was already present on master and it was probably the cause of some random terminal drawing bugs. Fixing it will significantly improve stability.